In [23]:
from glob import iglob
import os

import pandas as pd
import screed
import seaborn as sns
from tqdm import tqdm

# Change to Quest for Orthologs 2019 data directory

In [4]:
cd ~/data_sm/kmer-hashing/quest-for-orthologs/data/2019/

/mnt/data_sm/olga/kmer-hashing/quest-for-orthologs/data/2019


In [5]:
ls -lha

total 2.6G
drwxr-xr-x 5 olga root 4.0K Jan  8 16:08 ./
drwxr-xr-x 3 olga root 4.0K Dec 25 17:48 ../
drwxr-xr-x 5 olga czb  4.0K Dec 26 19:44 Archaea/
drwxr-xr-x 5 olga czb   16K Dec 26 19:44 Bacteria/
drwxr-xr-x 8 olga czb   32K Jan  8 08:13 Eukaryota/
-rw-r--r-- 1 olga czb   64M Jan  8 16:08 opisthokont_not_human_transcription_factors_ensembl_compara.csv
-rw-r--r-- 1 olga czb  2.6G Dec 25 18:46 QfO_release_2019_04.tar.gz
-rw-r--r-- 1 olga czb   18K May 10  2019 README
-rw-r--r-- 1 olga czb   12K Jan  8 07:47 species_metadata.csv


# Download orthology and transcription factor data

## Read orthologous transcription factors 

In [107]:
tfs_original = pd.read_csv('opisthokont_not_human_transcription_factors_ensembl_compara.csv')
print(tfs_original.shape)
tfs_original.head()

(343153, 16)


dn_ds    method_link_type       source__id  source__perc_id  \
0      NaN  ENSEMBL_PARALOGUES  ENSG00000267179          24.8792   
1      NaN  ENSEMBL_PARALOGUES  ENSG00000267179          25.6039   
2  0.35552  ENSEMBL_PARALOGUES  ENSG00000267179          26.3285   
3  0.33996  ENSEMBL_PARALOGUES  ENSG00000267179          27.2947   
4  0.18553  ENSEMBL_PARALOGUES  ENSG00000267179          20.2899   

   source__perc_pos source__protein_id source__species  source__taxon_id  \
0           37.1981    ENSP00000467286    homo_sapiens              9606   
1           37.1981    ENSP00000467286    homo_sapiens              9606   
2           38.1642    ENSP00000467286    homo_sapiens              9606   
3           38.8889    ENSP00000467286    homo_sapiens              9606   
4           25.8454    ENSP00000467286    homo_sapiens              9606   

        target__id  target__perc_id  target__perc_pos target__protein_id  \
0  ENSG00000175787          17.0813           25.5390    ENSP00000378792   
1  ENSG00000187815          20.1521           29.2776    ENSP00000361791   
2  ENSG00000172006          20.2602           29.3680    ENSP00000321132   
3  ENSG00000188629          23.4927           33.4719    ENSP00000415070   
4  ENSG00000270011          26.1682           33.3333    ENSP00000445323   

  target__species  target__taxon_id taxonomy_level           type  
0    homo_sapiens              9606      Bilateria  other_paralog  
1    homo_sapiens              9606      Bilateria  other_paralog  
2    homo_sapiens              9606      Bilateria  other_paralog  
3    homo_sapiens              9606      Bilateria  other_paralog  
4    homo_sapiens              9606      Bilateria  other_paralog

## Read random subset of TFs

In [106]:
tfs_random_subset = pd.read_csv('human_transcription_factors_with_uniprot_ids_random_subset100.csv')
print(tfs_random_subset.shape)
tfs_random_subset.head()

(100, 8)


Gene               ID               DBD  \
0  TFCP2L1  ENSG00000115112        Grainyhead   
1     PREB  ENSG00000138073           Unknown   
2     BPTF  ENSG00000171634           Unknown   
3     MXI1  ENSG00000119950              bHLH   
4    HNF4A  ENSG00000101076  Nuclear receptor   

  Motif status (Feb 2018)(Link to human TFs annotation)  \
0       Known motif – High-throughput in vitro [942]      
1  Likely sequence specific TF according to liter...      
2             Known motif – In vivo/Misc source [79]      
3            Known motif – In vivo/Misc source [553]      
4       Known motif – High-throughput in vitro [369]      

  IUPAC consensus (from selected PWM) uniprot_id  id_type            db_id  
0                    DCYRGHNNNDDCYRGH     Q9NZI6  Ensembl  ENSG00000115112  
1                                 NaN     H7C0W3  Ensembl  ENSG00000138073  
2                        KKKNTTGTKKNV     J3KSY0  Ensembl  ENSG00000171634  
3                            CCACGTGG     P50539  Ensembl  ENSG00000119950  
4                      VRGGTCAAAGTCCA     F8WBS7  Ensembl  ENSG00000101076

## Subset TF orthology to the random subset

In [108]:
tfs = tfs_original.query('source_id in @tfs_random_subset.db_id')
print(tfs.shape)
tfs.head()

UndefinedVariableError: name 'source_id' is not defined

# Go to Quest for Orthologs fastas

## Read species metadata

In [9]:
species_metadata = pd.read_csv("species_metadata.csv")
print(species_metadata.shape)
species_metadata.head()

(78, 10)


proteome_id  tax_id oscode  n_main_fasta_canonical  \
0  UP000007062    7165  ANOGA                   12553   
1  UP000000798  224324  AQUAE                    1553   
2  UP000006548    3702  ARATH                   27475   
3  UP000001570  224308  BACSU                    4260   
4  UP000001414  226186  BACTN                    4782   

   n_additional_fasta_isoforms  n_gene2acc  \
0                          971       13619   
1                            0        1557   
2                        14123       41920   
3                            7        4268   
4                            0        4823   

                                        species_name  \
0       Anopheles gambiae (African malaria mosquito)   
1                      Aquifex aeolicus (strain VF5)   
2             Arabidopsis thaliana (Mouse-ear cress)   
3                     Bacillus subtilis (strain 168)   
4  Bacteroides thetaiotaomicron (strain ATCC 2914...   

   divergence_from_human_mya               scientific_name  \
0                      797.0             Anopheles gambiae   
1                     4290.0              Aquifex aeolicus   
2                     1496.0          Arabidopsis thaliana   
3                     4290.0             Bacillus subtilis   
4                     4290.0  Bacteroides thetaiotaomicron   

                               common_name_or_strain  
0                           African malaria mosquito  
1                                         strain VF5  
2                                    Mouse-ear cress  
3                                         strain 168  
4  strain ATCC 29148 / DSM 2079 / NCTC 10582 / E5...

### Subset to opisthokonts

In [10]:
# Estimated opisthokonta divergence time from http://timetree.org/
t = 1105
opisthokonts = species_metadata.query('divergence_from_human_mya <= @t')
print(opisthokonts.shape)
opisthokonts.head()

(35, 10)


proteome_id  tax_id oscode  n_main_fasta_canonical  \
0  UP000007062    7165  ANOGA                   12553   
5  UP000007241  684364  BATDJ                    8610   
6  UP000009136    9913  BOVIN                   23774   
8  UP000001554    7739  BRAFL                   28542   
9  UP000001940    6239  CAEEL                   19986   

   n_additional_fasta_isoforms  n_gene2acc  \
0                          971       13619   
5                            0        8685   
6                        14534       38438   
8                            2       28608   
9                         8309       28507   

                                        species_name  \
0       Anopheles gambiae (African malaria mosquito)   
5  Batrachochytrium dendrobatidis (strain JAM81 /...   
6                                Bos taurus (Bovine)   
8  Branchiostoma floridae (Florida lancelet) (Amp...   
9                             Caenorhabditis elegans   

   divergence_from_human_mya                 scientific_name  \
0                      797.0               Anopheles gambiae   
5                     1105.0  Batrachochytrium dendrobatidis   
6                       96.0                      Bos taurus   
8                      684.0          Branchiostoma floridae   
9                      797.0          Caenorhabditis elegans   

       common_name_or_strain  
0   African malaria mosquito  
5  strain JAM81 / FGSC 10211  
6                     Bovine  
8           Florida lancelet  
9                        NaN

In [11]:
opisthokonts.query('scientific_name == "Homo sapiens"')

proteome_id  tax_id oscode  n_main_fasta_canonical  \
32  UP000005640    9606  HUMAN                   21146   

    n_additional_fasta_isoforms  n_gene2acc          species_name  \
32                        74769       96332  Homo sapiens (Human)   

    divergence_from_human_mya scientific_name common_name_or_strain  
32                        0.0    Homo sapiens                 Human

## Read Gene Accession file

```
Gene mapping files (*.gene2acc)
===============================

Column 1 is a unique gene symbol that is chosen with the following order of
preference from the annotation found in:
1) Model Organism Database (MOD)
2) Ensembl or Ensembl Genomes database
3) UniProt Ordered Locus Name (OLN)
4) UniProt Open Reading Frame (ORF)
5) UniProt Gene Name
A dash symbol ('-') is used when the gene encoding a protein is unknown.

Column 2 is the UniProtKB accession or isoform identifier for the given gene
symbol. This column may have redundancy when two or more genes have identical
translations.

Column 3 is the gene symbol of the canonical accession used to represent the
respective gene group and the first row of the sequence is the canonical one.
```

In [109]:

def read_gene2acc(gene2acc, names=['maybe_ensembl_id', 'uniprot_id', 'canonical_accession']):
    df = pd.read_csv(gene2acc, sep='\t', header=None, na_values='-', names=columns)
    return df

gene2acc = read_gene2acc('Eukaryota/UP000005640_9606.gene2acc')
# gene2acc = pd.read_csv('Eukaryota/UP000005640_9606.gene2acc', sep='\t', header=None, na_values='-', names=columns)
print(gene2acc.shape)
gene2acc.head()

(96332, 3)


maybe_ensembl_id  uniprot_id canonical_accession
0              NaN  A0A075B7B6                 NaN
1              NaN  A0A075B714                 NaN
2              NaN  A0A075B713                 NaN
3              NaN  A0A075B712                 NaN
4              NaN  A0A075B711                 NaN

In [29]:
gene2acc.dropna()

maybe_ensembl_id uniprot_id canonical_accession
600              ACOT7L     Q6ZUV0              ACOT7L
601              BARGIN     Q6ZT62              BARGIN
602              BARGIN   Q6ZT62-2              BARGIN
603                BCE1     O60756                BCE1
604            C1orf140     Q5VVS0            C1orf140
...                 ...        ...                 ...
96327  UNQ6190/PRO20217     Q6UXQ8    UNQ6190/PRO20217
96328  UNQ6493/PRO21345     Q6UXR8    UNQ6493/PRO21345
96329  UNQ6494/PRO21346     Q6UXR6    UNQ6494/PRO21346
96330  UNQ9165/PRO28630     Q6UXU0    UNQ9165/PRO28630
96331  UNQ9370/PRO34162     Q6UXP9    UNQ9370/PRO34162

[95732 rows x 3 columns]

## Read ID mapping file

```
Database mapping files (*.idmapping)
====================================

These files contain mappings from UniProtKB to other databases for each
reference proteome.
The format consists of three tab-separated columns:

1. UniProtKB accession
2. ID_type:
   Database name as shown in UniProtKB cross-references and supported by the ID
   mapping tool on the UniProt web site (http://www.uniprot.org/mapping)
3. ID:
   Identifier in the cross-referenced database.

```

In [17]:
opisthokonts.head()

proteome_id  tax_id oscode  n_main_fasta_canonical  \
0  UP000007062    7165  ANOGA                   12553   
5  UP000007241  684364  BATDJ                    8610   
6  UP000009136    9913  BOVIN                   23774   
8  UP000001554    7739  BRAFL                   28542   
9  UP000001940    6239  CAEEL                   19986   

   n_additional_fasta_isoforms  n_gene2acc  \
0                          971       13619   
5                            0        8685   
6                        14534       38438   
8                            2       28608   
9                         8309       28507   

                                        species_name  \
0       Anopheles gambiae (African malaria mosquito)   
5  Batrachochytrium dendrobatidis (strain JAM81 /...   
6                                Bos taurus (Bovine)   
8  Branchiostoma floridae (Florida lancelet) (Amp...   
9                             Caenorhabditis elegans   

   divergence_from_human_mya                 scientific_name  \
0                      797.0               Anopheles gambiae   
5                     1105.0  Batrachochytrium dendrobatidis   
6                       96.0                      Bos taurus   
8                      684.0          Branchiostoma floridae   
9                      797.0          Caenorhabditis elegans   

       common_name_or_strain  
0   African malaria mosquito  
5  strain JAM81 / FGSC 10211  
6                     Bovine  
8           Florida lancelet  
9                        NaN

In [34]:
opisthokonts.query('proteome_id == "UP000000437"')

proteome_id  tax_id oscode  n_main_fasta_canonical  \
17  UP000000437    7955  DANRE                   25939   

    n_additional_fasta_isoforms  n_gene2acc  \
17                        21219       47517   

                                   species_name  divergence_from_human_mya  \
17  Danio rerio (Zebrafish) (Brachydanio rerio)                      435.0   

   scientific_name common_name_or_strain  
17     Danio rerio             Zebrafish

# Get ID Mapping for uniprot ids from ENSMBL

In [37]:
dfs = []

for filename in tqdm(sorted(iglob("Eukaryota/*.idmapping"))):
#     print(filename)
    basename = os.path.basename(filename)
    prefix = basename.split('.')[0]
    species_id, taxa_id = prefix.split("_")
#     print(f"{species_id=} {taxa_id=}")
    
    if species_id in opisthokonts.proteome_id.values:

        df = pd.read_csv(filename, sep='\t', header=None, names=['uniprot_id', 'id_type', 'db_id'])
        df['species_id'] = species_id
        df['taxa_id'] = species_id

        # Use only Ensembl data
#         df = df.query('id_type == "Ensembl"')
        print(df.shape)
        dfs.append(df)
    
id_mapping = pd.concat(dfs, ignore_index=True)
print(id_mapping.shape)
id_mapping.head()

  2%|▏         | 1/48 [00:00<00:32,  1.44it/s]

(1036083, 5)


 10%|█         | 5/48 [00:01<00:18,  2.32it/s]

(555421, 5)
(149092, 5)
(141961, 5)


 12%|█▎        | 6/48 [00:02<00:25,  1.62it/s]

(1631036, 5)


 17%|█▋        | 8/48 [00:02<00:20,  1.96it/s]

(809657, 5)


 21%|██        | 10/48 [00:03<00:14,  2.54it/s]

(575653, 5)
(295804, 5)


 23%|██▎       | 11/48 [00:03<00:11,  3.12it/s]

(124774, 5)


 27%|██▋       | 13/48 [00:04<00:09,  3.64it/s]

(277055, 5)
(181143, 5)


 33%|███▎      | 16/48 [00:04<00:08,  3.86it/s]

(839553, 5)


 40%|███▉      | 19/48 [00:05<00:06,  4.41it/s]

(565651, 5)


 42%|████▏     | 20/48 [00:05<00:08,  3.43it/s]

(549426, 5)


 46%|████▌     | 22/48 [00:06<00:07,  3.42it/s]

(518273, 5)
(205999, 5)


 50%|█████     | 24/48 [00:06<00:07,  3.35it/s]

(765068, 5)
(146229, 5)


 54%|█████▍    | 26/48 [00:07<00:06,  3.56it/s]

(594466, 5)


 56%|█████▋    | 27/48 [00:08<00:09,  2.18it/s]

(1143307, 5)


 58%|█████▊    | 28/48 [00:08<00:08,  2.28it/s]

(424236, 5)


 62%|██████▎   | 30/48 [00:09<00:05,  3.15it/s]

(268306, 5)
(149870, 5)


 65%|██████▍   | 31/48 [00:09<00:07,  2.32it/s]

(873723, 5)


 67%|██████▋   | 32/48 [00:09<00:05,  2.76it/s]

(214092, 5)


 71%|███████   | 34/48 [00:11<00:06,  2.04it/s]

(2668934, 5)


 81%|████████▏ | 39/48 [00:12<00:02,  3.24it/s]

(340838, 5)
(151313, 5)


 83%|████████▎ | 40/48 [00:12<00:02,  3.24it/s]

(315939, 5)


 88%|████████▊ | 42/48 [00:12<00:01,  3.44it/s]

(637633, 5)


 90%|████████▉ | 43/48 [00:13<00:01,  3.49it/s]

(281658, 5)


 92%|█████████▏| 44/48 [00:13<00:01,  3.46it/s]

(307893, 5)


 94%|█████████▍| 45/48 [00:14<00:01,  2.59it/s]

(773011, 5)


 96%|█████████▌| 46/48 [00:14<00:00,  2.45it/s]

(561122, 5)


100%|██████████| 48/48 [00:14<00:00,  3.24it/s]

(390075, 5)


(19464294, 5)


uniprot_id       id_type        db_id   species_id      taxa_id
0     Q5PRD0  UniProtKB-ID  143BA_DANRE  UP000000437  UP000000437
1     Q5PRD0     Gene_Name       ywhaba  UP000000437  UP000000437
2     Q5PRD0  Gene_Synonym       ywhab1  UP000000437  UP000000437
3     Q5PRD0  Gene_ORFName   wu:fb80c08  UP000000437  UP000000437
4     Q5PRD0            GI     82592598  UP000000437  UP000000437

# Merge id mapping with ensembl compara tfs

In [40]:
tfs.tail()

dn_ds     method_link_type       source__id  source__perc_id  \
343148    NaN  ENSEMBL_ORTHOLOGUES  ENSG00000036549         36.43410   
343149    NaN  ENSEMBL_ORTHOLOGUES  ENSG00000036549         26.91030   
343150    NaN  ENSEMBL_ORTHOLOGUES  ENSG00000036549         28.90370   
343151    NaN  ENSEMBL_ORTHOLOGUES  ENSG00000036549          7.08749   
343152    NaN  ENSEMBL_ORTHOLOGUES  ENSG00000036549         13.84270   

        source__perc_pos source__protein_id source__species  source__taxon_id  \
343148          52.27020    ENSP00000359837    homo_sapiens              9606   
343149          33.44410    ENSP00000359837    homo_sapiens              9606   
343150          45.29350    ENSP00000359837    homo_sapiens              9606   
343151           9.30233    ENSP00000359837    homo_sapiens              9606   
343152          21.04100    ENSP00000359837    homo_sapiens              9606   

                 target__id  target__perc_id  target__perc_pos  \
343148   ENSEBUG00000002277          39.2133           56.2575   
343149   ENSPMAG00000000497          56.6434           70.3963   
343150   ENSCING00000002486          25.1203           39.3648   
343151  ENSCSAVG00000000992          48.4848           63.6364   
343152          FBgn0031876          35.1124           53.3708   

         target__protein_id          target__species  target__taxon_id  \
343148   ENSEBUP00000003097       eptatretus_burgeri              7764   
343149   ENSPMAP00000000552       petromyzon_marinus              7757   
343150   ENSCINP00000005056       ciona_intestinalis              7719   
343151  ENSCSAVP00000001713           ciona_savignyi             51511   
343152          FBpp0078990  drosophila_melanogaster              7227   

       taxonomy_level              type  
343148     Vertebrata  ortholog_one2one  
343149     Vertebrata  ortholog_one2one  
343150       Chordata  ortholog_one2one  
343151       Chordata  ortholog_one2one  
343152      Bilateria  ortholog_one2one

In [76]:
id_mapping_for_merging = id_mapping.copy()
id_mapping_for_merging.columns = "target__" + id_mapping_for_merging.columns
id_mapping_for_merging.head()

target__uniprot_id target__id_type target__db_id target__species_id  \
0             Q5PRD0    UniProtKB-ID   143BA_DANRE        UP000000437   
1             Q5PRD0       Gene_Name        ywhaba        UP000000437   
2             Q5PRD0    Gene_Synonym        ywhab1        UP000000437   
3             Q5PRD0    Gene_ORFName    wu:fb80c08        UP000000437   
4             Q5PRD0              GI      82592598        UP000000437   

  target__taxa_id  
0     UP000000437  
1     UP000000437  
2     UP000000437  
3     UP000000437  
4     UP000000437

In [78]:
tfs.shape

(343153, 16)

In [79]:
%%time

tfs_uniprot_merge_proteins = tfs.merge(id_mapping_for_merging, left_on='target__protein_id', right_on='target__db_id')
print(tfs_uniprot_merge_proteins.shape)
tfs_uniprot_merge_proteins.head()

(59232, 21)
CPU times: user 20 s, sys: 699 ms, total: 20.7 s
Wall time: 20.8 s


dn_ds    method_link_type       source__id  source__perc_id  \
0      NaN  ENSEMBL_PARALOGUES  ENSG00000267179          24.8792   
1  0.37910  ENSEMBL_PARALOGUES  ENSG00000181666          45.6753   
2  0.43907  ENSEMBL_PARALOGUES  ENSG00000164256          26.1745   
3      NaN  ENSEMBL_PARALOGUES  ENSG00000136866          30.5426   
4      NaN  ENSEMBL_PARALOGUES  ENSG00000187815          29.6578   

   source__perc_pos source__protein_id source__species  source__taxon_id  \
0           37.1981    ENSP00000467286    homo_sapiens              9606   
1           57.5114    ENSP00000315505    homo_sapiens              9606   
2           34.0045    ENSP00000296682    homo_sapiens              9606   
3           41.7054    ENSP00000452552    homo_sapiens              9606   
4           41.4449    ENSP00000361791    homo_sapiens              9606   

        target__id  target__perc_id  ...  target__protein_id target__species  \
0  ENSG00000175787          17.0813  ...     ENSP00000378792    homo_sapiens   
1  ENSG00000175787          49.9171  ...     ENSP00000378792    homo_sapiens   
2  ENSG00000175787          38.8060  ...     ENSP00000378792    homo_sapiens   
3  ENSG00000175787          32.6700  ...     ENSP00000378792    homo_sapiens   
4  ENSG00000175787          25.8706  ...     ENSP00000378792    homo_sapiens   

  target__taxon_id  taxonomy_level           type target__uniprot_id  \
0             9606       Bilateria  other_paralog             Q14929   
1             9606       Bilateria  other_paralog             Q14929   
2             9606       Bilateria  other_paralog             Q14929   
3             9606       Bilateria  other_paralog             Q14929   
4             9606       Bilateria  other_paralog             Q14929   

  target__id_type    target__db_id target__species_id target__taxa_id  
0     Ensembl_PRO  ENSP00000378792        UP000005640     UP000005640  
1     Ensembl_PRO  ENSP00000378792        UP000005640     UP000005640  
2     Ensembl_PRO  ENSP00000378792        UP000005640     UP000005640  
3     Ensembl_PRO  ENSP00000378792        UP000005640     UP000005640  
4     Ensembl_PRO  ENSP00000378792        UP000005640     UP000005640  

[5 rows x 21 columns]

In [80]:
%%time

tfs_uniprot_merge_genes = tfs.merge(id_mapping_for_merging, left_on='target__id', right_on='target__db_id')
print(tfs_uniprot_merge_genes.shape)
tfs_uniprot_merge_genes.head()

(266814, 21)
CPU times: user 19.6 s, sys: 757 ms, total: 20.4 s
Wall time: 20.3 s


dn_ds    method_link_type       source__id  source__perc_id  \
0     NaN  ENSEMBL_PARALOGUES  ENSG00000267179          24.8792   
1     NaN  ENSEMBL_PARALOGUES  ENSG00000267179          24.8792   
2     NaN  ENSEMBL_PARALOGUES  ENSG00000267179          24.8792   
3     NaN  ENSEMBL_PARALOGUES  ENSG00000267179          24.8792   
4  0.3791  ENSEMBL_PARALOGUES  ENSG00000181666          45.6753   

   source__perc_pos source__protein_id source__species  source__taxon_id  \
0           37.1981    ENSP00000467286    homo_sapiens              9606   
1           37.1981    ENSP00000467286    homo_sapiens              9606   
2           37.1981    ENSP00000467286    homo_sapiens              9606   
3           37.1981    ENSP00000467286    homo_sapiens              9606   
4           57.5114    ENSP00000315505    homo_sapiens              9606   

        target__id  target__perc_id  ...  target__protein_id target__species  \
0  ENSG00000175787          17.0813  ...     ENSP00000378792    homo_sapiens   
1  ENSG00000175787          17.0813  ...     ENSP00000378792    homo_sapiens   
2  ENSG00000175787          17.0813  ...     ENSP00000378792    homo_sapiens   
3  ENSG00000175787          17.0813  ...     ENSP00000378792    homo_sapiens   
4  ENSG00000175787          49.9171  ...     ENSP00000378792    homo_sapiens   

  target__taxon_id  taxonomy_level           type target__uniprot_id  \
0             9606       Bilateria  other_paralog             Q7Z761   
1             9606       Bilateria  other_paralog             Q14929   
2             9606       Bilateria  other_paralog             Q5SR55   
3             9606       Bilateria  other_paralog             S4R440   
4             9606       Bilateria  other_paralog             Q7Z761   

  target__id_type    target__db_id target__species_id target__taxa_id  
0         Ensembl  ENSG00000175787        UP000005640     UP000005640  
1         Ensembl  ENSG00000175787        UP000005640     UP000005640  
2         Ensembl  ENSG00000175787        UP000005640     UP000005640  
3         Ensembl  ENSG00000175787        UP000005640     UP000005640  
4         Ensembl  ENSG00000175787        UP000005640     UP000005640  

[5 rows x 21 columns]

## Add human uniprot ids

In [84]:
human_id_mapping = pd.read_csv('Eukaryota/UP000005640_9606.idmapping', sep='\t', header=None, names=['uniprot_id', 'id_type', 'db_id'])
human_id_mapping.columns = 'source__' + human_id_mapping.columns
print(human_id_mapping.shape)
human_id_mapping.head()

(2668934, 3)


source__uniprot_id source__id_type    source__db_id
0           P62258-2         UniParc    UPI00001E6021
1           P62258-2     Ensembl_TRS  ENST00000571732
2           P62258-2     Ensembl_PRO  ENSP00000461762
3           P62258-2     Ensembl_TRS  ENST00000616643
4           P62258-2     Ensembl_PRO  ENSP00000481059

In [102]:
%%time 
tfs_uniprot_merge_proteins_with_human = tfs_uniprot_merge_proteins.merge(
    human_id_mapping, left_on='source__protein_id', right_on='source__db_id', how='outer')
tfs_uniprot_merge_proteins_with_human.columns = tfs_uniprot_merge_proteins_with_human.columns.str.replace("source__", 'human__')
print(tfs_uniprot_merge_proteins_with_human.shape)
tfs_uniprot_merge_proteins_with_human.query('type == "ortholog_one2one"').head()

(2727198, 24)
CPU times: user 2.24 s, sys: 475 ms, total: 2.71 s
Wall time: 2.76 s


dn_ds     method_link_type        human__id  human__perc_id  \
99   0.86154  ENSEMBL_ORTHOLOGUES  ENSG00000181666         97.4203   
100  0.57895  ENSEMBL_ORTHOLOGUES  ENSG00000181666         93.6267   
148  0.91702  ENSEMBL_ORTHOLOGUES  ENSG00000164256         85.5705   
149  0.77097  ENSEMBL_ORTHOLOGUES  ENSG00000164256         77.5168   
199  2.72000  ENSEMBL_ORTHOLOGUES  ENSG00000136866         96.2791   

     human__perc_pos human__protein_id human__species  human__taxon_id  \
99           98.1791   ENSP00000315505   homo_sapiens           9606.0   
100          94.3854   ENSP00000315505   homo_sapiens           9606.0   
148          87.0246   ENSP00000296682   homo_sapiens           9606.0   
149          78.5235   ENSP00000296682   homo_sapiens           9606.0   
199          96.7442   ENSP00000452552   homo_sapiens           9606.0   

             target__id  target__perc_id  ...  taxonomy_level  \
99   ENSGGOG00000013141          97.5684  ...       Homininae   
100  ENSPTRG00000010907          95.2160  ...       Homininae   
148  ENSGGOG00000024584          93.0657  ...       Homininae   
149  ENSPTRG00000028674          95.5862  ...       Homininae   
199  ENSGGOG00000000731          98.4152  ...       Homininae   

                 type target__uniprot_id  target__id_type       target__db_id  \
99   ortholog_one2one         A0A2I2Z547      Ensembl_PRO  ENSGGOP00000042196   
100  ortholog_one2one             H2R440      Ensembl_PRO  ENSPTRP00000046527   
148  ortholog_one2one             G3S233      Ensembl_PRO  ENSGGOP00000022130   
149  ortholog_one2one             H2R5H8      Ensembl_PRO  ENSPTRP00000048166   
199  ortholog_one2one             G3QEM3      Ensembl_PRO  ENSGGOP00000000720   

    target__species_id target__taxa_id human__uniprot_id human__id_type  \
99         UP000001519     UP000001519               NaN            NaN   
100        UP000002277     UP000002277               NaN            NaN   
148        UP000001519     UP000001519            Q9NQV7    Ensembl_PRO   
149        UP000002277     UP000002277            Q9NQV7    Ensembl_PRO   
199        UP000001519     UP000001519          Q9Y6Q3-2    Ensembl_PRO   

        human__db_id  
99               NaN  
100              NaN  
148  ENSP00000296682  
149  ENSP00000296682  
199  ENSP00000452552  

[5 rows x 24 columns]

In [103]:
tfs_uniprot_merge_proteins_with_human.type.value_counts()

other_paralog             39292
ortholog_one2one          11503
ortholog_many2many         4898
ortholog_one2many          2421
within_species_paralog     1067
alt_allele                   73
Name: type, dtype: int64

## Write merged TFs to disk

In [104]:
pwd

'/mnt/data_sm/olga/kmer-hashing/quest-for-orthologs/data/2019'

In [105]:
%time tfs_uniprot_merge_proteins_with_human.to_csv('opisthokont_not_human_transcription_factors_ensembl_compara_merged_uniprot.csv.gz', index=False)
%time tfs_uniprot_merge_proteins_with_human.to_parquet('opisthokont_not_human_transcription_factors_ensembl_compara_merged_uniprot.parquet', index=False)

CPU times: user 28.9 s, sys: 137 ms, total: 29.1 s
Wall time: 29.3 s
CPU times: user 2.54 s, sys: 241 ms, total: 2.78 s
Wall time: 2.88 s


## Make set variable for quick membership evalution

In [93]:
tf_orthologs = set(tfs_uniprot_merge_proteins_with_human.target__uniprot_id)

### Prove that the set `tf_orthologs` is faster

In [95]:
%timeit 'Q7Z761' in tf_orthologs

68.2 ns ± 0.876 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [96]:
%timeit 'Q7Z761' in tfs_uniprot_merge_proteins_with_human.target__uniprot_id

14.1 µs ± 236 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


#### Yep, sets are 3 orders of magnitude faster!

# Read non-human proteins and subset if they are an ortholog of a TF

## Make outdir

In [43]:
not_human_outdir = 'Eukaryota/not-human-transcription-factor-fastas/'
! mkdir $not_human_outdir


mkdir: cannot create directory ‘Eukaryota/not-human-transcription-factor-fastas/’: File exists


## How much compute is this?

Number of human transcription factor proteins in the quest for orthologs database


In [55]:
n_human_qfo_tfs = 5749

In [50]:
tfs_uniprot.query('target__species != "homo_sapiens"').shape

(48947, 21)

In [53]:
n_not_human_qfo_tfs = 48947

In [57]:
n_human_qfo_tfs * n_not_human_qfo_tfs * 0.0006 / 60 / 60

46.89938383333333


### Okay, 46 hours is reasonable!

## Read in protein fastas with screed

In [61]:


for filename in iglob('Eukaryota/not-human-protein-fastas/*.fasta'):
    tf_records = []

    basename = os.path.basename(filename)
    with screed.open(filename) as records:
        for record in records:
            name = record['name']
            record_id = name.split()[0]
            uniprot_id = record_id.split('|')[1]
            if uniprot_id in tf_orthologs:
                tf_records.append(record)
                
    if len(tf_records) > 0:
        print(filename)
        print(f"\tlen(tf_records): {len(tf_records)}")
        with open(f'{not_human_outdir}/{basename}', 'w') as f:
            for record in tf_records:
                f.write(">{name}\n{sequence}\n".format(**record))


Eukaryota/not-human-protein-fastas/UP000000437_7955_additional.fasta
	len(tf_records): 1171
Eukaryota/not-human-protein-fastas/UP000008143_8364.fasta
	len(tf_records): 493
Eukaryota/not-human-protein-fastas/UP000001038_8090.fasta
	len(tf_records): 1263
Eukaryota/not-human-protein-fastas/UP000002280_13616.fasta
	len(tf_records): 1007
Eukaryota/not-human-protein-fastas/UP000000539_9031_additional.fasta
	len(tf_records): 640
Eukaryota/not-human-protein-fastas/UP000000803_7227_additional.fasta
	len(tf_records): 404
Eukaryota/not-human-protein-fastas/UP000018468_7918.fasta
	len(tf_records): 979
Eukaryota/not-human-protein-fastas/UP000002254_9615.fasta
	len(tf_records): 1231
Eukaryota/not-human-protein-fastas/UP000001940_6239.fasta
	len(tf_records): 544
Eukaryota/not-human-protein-fastas/UP000000437_7955.fasta
	len(tf_records): 1156
Eukaryota/not-human-protein-fastas/UP000002311_559292.fasta
	len(tf_records): 48
Eukaryota/not-human-protein-fastas/UP000000589_10090.fasta
	len(tf_records): 124

# Script to run

In [62]:
ll /mnt/data_sm/olga/kmer-hashing/quest-for-orthologs/data/2019/Eukaryota/human-transcription-factor-fastas/

total 2520
-rw-r--r-- 1 olga 2575632 Jan  8 08:10 human_transcription_factor_proteins.fasta


In [64]:
%%file qfo_human_vs_opisthokont_tfs.sh
#!/bin/bash
OUTDIR=$HOME/data_sm/kmer-hashing/quest-for-orthologs/analysis/2019/transcription-factors/
mkdir -p $OUTDIR/intermediates
cd $OUTDIR/intermediates

PARQUET=$OUTDIR/qfo-eukaryota-tfs-protein.parquet

EUKARYOTA=/mnt/data_sm/olga/kmer-hashing/quest-for-orthologs/data/2019/Eukaryota
HUMAN=$EUKARYOTA/human-transcription-factor-fastas/human_transcription_factor_proteins.fasta
NOT_HUMAN=$EUKARYOTA/not-human-transcription-factor-fastas/

conda activate khtools--encodings--compare-cli


time khtools compare-kmers \
    --processes 120 \
    --ksize-min 3 \
    --ksize-max 45 \
    --parquet $PARQUET \
    --intermediate-parquet \
    --fastas2 $HUMAN \
    $NOT_HUMAN/* | tee khtools_compare-kmers.log

Writing qfo_human_vs_opisthokont_tfs.sh


In [65]:
pwd

'/mnt/data_sm/olga/kmer-hashing/quest-for-orthologs/data/2019'

## Time estimation

taking ~1000 seconds per non-human sequence

In [70]:
n_not_human_qfo_tfs

48947

In [69]:
n_not_human_qfo_tfs * 1000 / 120 / 60 / 60 / 24

4.720968364197531

Okay, so this will take ~4.7 days to compute running on `lrrr`